### DISCUSSION TITLE
MicronautStreamHandler closes MongoClient after processing request which is undesirable

### DISCUSSION THREAD

**Discussion by mayurtanna** [micronaut-projects/micronaut-aws#1177] [General]
Hi @sdelamo 

Like I mentioned in the issue https://github.com/micronaut-projects/micronaut-aws/issues/1077, when I use MicronautRequestStreamHandler, it seems to be destroying the MongoClient (so ultimately connectionpool) once processing the request and hence we get an exception like below for subsequent request.

So, what I believe is it might be due to StreamFunctionExecutor class's execute method which is getting called internally when we use MicronautRequestStreamHandler. In that method it is closing the applicationContext in finally clause which seem to be destroying MongoClient internally.

I would like to understand, is it desirable to close the applicationContext post handling the request ? Because in case of MicronautRequestHandler I didn't find like we are closing applicationContext post handling of the request. Also, what should I do to keep the connection open ? 

Can you provide your thoughts here ?

----------------Exception---------------
Connection pool shut down: java.lang.IllegalStateException
java.lang.IllegalStateException: Connection pool shut down
at org.apache.http.util.Asserts.check(Asserts.java:34)
at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.requestConnection(PoolingHttpClientConnectionManager.java:269)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
at java.base/java.lang.reflect.Method.invoke(Unknown Source)
at software.amazon.awssdk.http.apache.internal.conn.ClientConnectionManagerFactory$Handler.invoke(ClientConnectionManagerFactory.java:8

**Comment by mayurtanna** [comment]
@sdelamo  did you get a chance to look at this ?

**Reply by sdelamo** [reply]
Not yet. I will attempt to look into to on Friday.

**Reply by graemerocher** [reply]
A bug report with a reproducer would be nice

**Reply by mayurtanna** [reply]
@sdelamo  Did you get a chance to look at it ? @graemerocher  I understand a reproducer will be helpful but at least from code @sdelamo  can understand what I am trying to explain. If I am not wrong, we are building and destroying applicationContext with each request in case of  MicronautRequestStreamHandler which is something not desirable. So, he can give some comments / insights on it. Is it a design flow ? or I am understanding it differently. I appreciate if you @sdelamo can check it quickly.

**Reply by sdelamo** [reply]
I've moved this into an issue:  

https://github.com/micronaut-projects/micronaut-aws/issues/1187

and I created a PR with a test

https://github.com/micronaut-projects/micronaut-aws/pull/1188

**Reply by mayurtanna** [reply]
@sdelamo  Thank you.
